### PR TITLE
Reduce CPU from source-control git polling on Windows

### LINKED ISSUE #6381: [Bug]: Source-control polling spawns recurring git.exe bursts → periodic CPU — Windows, v1.4.97
### Operating system

Windows

### Orca version

v1.4.97

### Details

**Summary**

On Windows, Orca v1.4.97 emits recurring bursts of short-lived `git.exe` processes for workspace source-control refresh while Orca stays open, with no manual git action on my part. Each spawned `git.exe` is scanned by Windows Defender real-time protection, so the churn surfaces as periodic CPU on both Orca's process tree and `MsMpEng.exe`.

This is distinct from the lightweight ~3s git-status badge poll — these are the heavier refresh set (`rev-parse @{upstream}`, `remote get-url`, `config`, `status --porcelain`) fired across the open repos and not coalesced.

**What I measured** (Windows 10 Pro 19045, 8-core; 6 repos / 6 projects open)

Spawns are parented by `Orca.exe`, running `git status --porcelain` / `git config` / `git rev-parse --abbrev-ref --symbolic-full-name <branch>@{upstream}` / `git remote get-url origin`. The bursts are intermittent — they cluster during active use (in one window I captured ~6 bursts over ~6 minutes, peaking ~10 concurrent `git.exe`; a later 6-minute window showed 9 `Orca.exe`-parented spawns peaking at 6). During bursts, Windows Defender (`MsMpEng.exe`) rises in lockstep — near-0 when idle, up to ~15–22% of a single core (≈2–3% of total CPU on this 8-core box) as it scans each spawned process. (Measured via `Win32_Process` `KernelModeTime+UserModeTime`, since `(Get-Process MsMpEng).CPU` returns null for the PPL-protected process.)

**Reproduce** — paste in PowerShell while Orca is open with a few git repos. Run it for several minutes (the bursts are intermittent, so a 1-minute window can miss them):

```powershell
$seen=@{}; $peak=0; $end=(Get-Date).AddMinutes(10)
while((Get-Date) -lt $end){
  $g=@(Get-CimInstance Win32_Process -Filter "Name='git.exe'")
  if($g.Count -gt $peak){ $peak=$g.Count }
  if($g.Count){ $names=@{}; foreach($p in Get-Process){ $names[$p.Id]=$p.ProcessName }
    foreach($x in $g){ if(-not $seen.ContainsKey([int]$x.ProcessId)){
      $par = if($names.ContainsKey([int]$x.ParentProcessId)){ $names[[int]$x.ParentProcessId] } else { '?' }
      $seen[[int]$x.ProcessId] = "$par | $($x.CommandLine)" } } }
  Start-Sleep -Milliseconds 250
}
"git.exe spawns: $($seen.Count)   peak concurrent: $peak"
$seen.Values | Group-Object | Sort-Object Count -Descending | Select-Object -First 20 |
  ForEach-Object{ '{0,4}  {1}' -f $_.Count, $_.Name }
```

Expected: a handful of git invocations. Observed: recurring bursts, parented by `Orca.exe`, doing status/config/rev-parse/remote across the open repos.

**Expected:** background source-control refresh is throttled/coalesced so it doesn't spawn repeated git-process bursts for as long as Orca is open.

**Actual:** recurring `git.exe` spawn bursts during use, driving periodic CPU on Orca + Defender (which scans every spawn).

**Context — this area is being actively tuned, but it persists:**
- v1.4.94 changelog: "Reduce repeated git probes during worktree refresh"
- v1.4.95 changelog: "Reduce source-control background load during refreshes"
- Still reproduces on v1.4.97.

**Not a duplicate of:**
- #5386 (closed as completed/fixed, macOS, v1.4.68) — "high idle CPU with many workspaces." That was fixed pre-1.4.94; I'm on v1.4.97 *after* the source-control-load tuning, on Windows, and it still reproduces with Defender amplification.
- #5381 (closed `not_planned`, macOS, v1.4.68) — a pathological `git worktree` reconcile storm (~25/sec change-broadcasts / ~10/sec git subprocesses) caused by a stale/unresolvable gitdir pointer. My case is healthy repos doing steady-state polling, with a different command set (`status` / `config` / `rev-parse upstream` / `remote get-url`, not `git worktree`).

**Caveats / scope:**
- Reproduces with several repos open; the per-repo refresh mechanism suggests the cost scales with the number of open repos (i.e. the refresh isn't coalesced across workspaces). Happy to run a single-repo comparison if that would help.
- I have **not** A/B-tested against an older build, so I'm not pinpointing a regression version — only reporting current v1.4.97 behavior.
- Scope: measurements isolate `git.exe` parented by `Orca.exe`; other tools on the box also spawn git and are excluded by parent-PID filtering. Defender amplifies the cost, but the root here is Orca's spawn cadence — the same churn would tax any on-access AV.

### PR DESCRIPTION
**Pull request by brennanb2025**
## What changes

Source-control polling's background upstream check used to spawn a burst of short-lived `git` processes each time it ran for a branch with no upstream — up to ~8 separate `git config --get` calls per probe round (plus the `rev-parse`/`symbolic-ref` reads). On Windows, every spawn is scanned by Windows Defender, so the churn surfaced as periodic CPU. This collapses all of a probe round's `git config --get` reads into **one** `git config --list -z` snapshot, dropping a no-upstream probe round from ~10 to ~4 `git` processes.

## What does NOT change

No change to **how often** the probe runs (the 3 s poll cadence and the 5-minute no-upstream negative cache are untouched), no change to caching/staleness, and **no change to the resolved upstream or push-target** — only the number of `git` subprocesses spawned per round. `git remote` / `git remote get-url` are intentionally left as real-git calls (they are not on the steady-state polling path).

Closes #6381.

## Background

This is the residual tail of the source-control git-burst work. Prior releases (v1.4.94–v1.4.101 and the upstream/push-target fix that followed) already eliminated the recurring bursts; the reporter confirmed on v1.4.103 that peak concurrent git dropped from ~10 to 3 with no sustained bursts, leaving only a low-rate `config` / `rev-parse @{u}` probe set. This change targets that remaining per-round fan-out so each residual refresh spawns materially fewer processes — which is what the on-access AV scanning cost scales with.

## Design approach

A small `runGit` wrapper, `createGitConfigSnapshotRunner`, intercepts only `git config --get <key>` calls. On the first such call in a round it lazily fetches `git config --list -z` once (single-flighting the in-flight promise so the resolvers' concurrent `Promise.all` reads share one fetch), parses it into a key → ordered-values map, and serves every subsequent `config --get` from that snapshot. Lookups apply git's key canonicalization (lowercase the section and final key segment, preserve the subsection) and multivar `--get` last-wins semantics, so a served value is byte-identical to what real `git config --get` would return. If the snapshot fetch fails for any reason, interception is disabled for the round and calls fall back to real git, preserving today's behavior exactly. The wrapper is created fresh per probe round, so there is no cross-round config caching (staleness is unchanged).

Wired into the two polling probe sites that drive the residual churn: `probeEffectiveUpstreamStatus` (local, `src/main/git/status.ts`) and `readOrProbeNoEffectiveUpstreamStatus` (SSH relay, `src/relay/git-status-upstream-negative-cache.ts`). The same-name-origin `rev-parse` detection stays on the outer runner in both, so the negative-cache write gate is unaffected.

## Design-review outcome

Ran an iterative design review on the doc. It caught a real correctness bug in an earlier draft (a multivar value would have been resolved first-wins when `git config --get` is actually last-wins) and a single-flight gap (memoizing the parsed value instead of the in-flight promise would have let concurrent first-calls each spawn `config --list`); both were fixed before implementation. The review also flagged that synthesizing `git remote` output in the wrapper re-implemented too much git behavior with silent wrong-resolution risk — so the final scope was narrowed to intercept only `config --get` and leave `remote`/`remote get-url` as real-git pass-through, which keeps the full per-round saving for the common case while removing the riskiest surface.

## Implementation & deviations

Implemented as designed. One in-scope addition beyond the listed files: `src/main/git/status-upstream-negative-cache.test.ts` mocks needed to handle the new `config --list -z` subprocess shape, so it was updated alongside the churn tests. No other deviations.

## Completeness verification

A read-only verification pass mapped every design requirement, edge case, and validation item to a PASS with file:line evidence, and empirically confirmed against real git that the served values match for the queried key shapes (including the `feature/fix` slash subsection). Verdict: complete, no behavior change beyond subprocess-count reduction.

## Headline behavior status

**Implemented** in production code on the real polling path (both local and relay probe sites execute the coalescing — not behind a flag or fixture). The deterministic churn tests assert the spawn-count reduction on the production path.

## Perf audit

Touched hot paths: the 3 s source-control poll's upstream probe and its SSH-relay equivalent. The audit confirmed the change strictly reduces spawn count with no new per-poll cost (the wrapper only instantiates inside a probe round, which is itself gated by the negative cache), no leak (per-round closure, fallback `.catch` observed so no unhandled rejection), and correct composition with the existing in-flight dedup / negative cache. A large-config timing check (~88 KB, 2300+ entries) showed `git config --list -z` runs in ~0.05 s — process spawn dominates, so one `--list` is cheaper than the several `--get`s it replaces even on big configs. No measurement gap: the deterministic "exactly one `config --list`, zero `config --get` per round" tests are the correct proxy for the churn-reduction claim.

## Code-review loop

Two independent reviewers in one round (one Codex, one Claude), both adversarially targeting any case where a snapshot-served value could diverge from real `git config --get`. Both returned clean — no P0/P1/P2 actionable findings. Each independently validated the parser, key canonicalization, single-flight, fallback, and the `.trim()` consumer equivalence against the real git binary.

## Validation

- `pnpm run typecheck` — pass.
- `pnpm exec oxlint` (repo-wide) — 0 errors (one pre-existing unrelated warning in `useGitStatusPolling.ts`, not modified here).
- Focused: `git-config-snapshot-runner.test.ts`, `status-upstream-probe-churn.test.ts`, `status-upstream-negative-cache.test.ts`, relay `git-status-upstream-negative-cache.test.ts` — 29 pass.
- Broader git suites (`src/main/git/` + `git-effective-upstream` + `git-configured-branch-target`) — 384 pass.
- Real-git integration check (scratch, not shipped): snapshot runner byte-identical to real `git config --get` across mixed-case / slash-subsection / dotted-name / multivar / missing keys; `getEffectiveGitUpstreamStatus` resolves identically with vs without the wrapper while spawning 1 `config --list` + 0 `config --get`.

## Cross-platform & SSH

Pure shift in git argument shape behind `runGit` — no OS-specific code, no path assumptions, no shell. The reduction is real on macOS/Linux too; it's largest on Windows where each spawn carries the Defender-scan cost. The SSH relay probe path is wired with the same wrapper, so remote worktrees get the same one-for-one reduction.

## Accepted gaps

The actual Windows-Defender CPU drop can't be observed from a macOS dev box; this issue is Windows-specific. The root cause the reporter identified is spawn cadence ("the same churn would tax any on-access AV"), and that reduction is proven deterministically and against real git.

Made with [Orca](https://github.com/stablyai/orca) 🐋

## Post-PR stabilization

- CI: `verify` and `track-community-pr` pass.
- CodeRabbit raised two findings on the config-snapshot runner, both addressed in 56dedc96e and acknowledged by CodeRabbit: (1) Major — an absent key now throws a typed `GitConfigSnapshotKeyNotFoundError` to match `git config --get`'s exit-1 contract (without re-spawning real-git, which would defeat the coalescing); (2) Minor — added a loaded-snapshot missing-key regression test. No remaining actionable comments.

### CODE REVIEW

**Comment by brennanb2025** [comment]
@coderabbitai review

**Comment by brennanb2025** [comment]
@coderabbitai review

File: src/shared/git-config-snapshot-runner.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    const configSnapshot = await readSnapshot()
+    if (!configSnapshot) {
+      return runGit(args)
+    }
+
+    const values = configSnapshot.get(canonicalizeGitConfigLookupKey(args[2] ?? ''))
+    if (!values?.length) {
+      return { stdout: '' }
+    }
```
[the remark this answers is not in the document]
Comment (brennanb2025) [reply src/shared/git-config-snapshot-runner.ts:98]: Addressed in 56dedc96e. Adopted the second option you flagged: a snapshot miss now throws a typed `GitConfigSnapshotKeyNotFoundError` rather than resolving `{ stdout: '' }`, so the wrapper preserves `git config --get`'s exit-1-on-absent-key contract and stays a faithful drop-in. I deliberately did *not* delegate misses to the real runner — on a no-upstream branch most queried keys are absent, so a passthrough would re-spawn the exact `config --get` subprocesses this change exists to coalesce. The sole probe-path consumer (`getGitConfigValue`) already catches and maps to null, so resolution is unchanged (verified against the real git binary).

File: src/shared/git-config-snapshot-runner.test.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    await expect(runner(['config', '--get', 'branch.main.remote'])).resolves.toEqual({
+      stdout: 'real:branch.main.remote'
+    })
+    await expect(runner(['config', '--get', 'branch.main.merge'])).resolves.toEqual({
+      stdout: 'real:branch.main.merge'
+    })
+
+    expect(countCalls(runGit, ['config', '--list', '-z'])).toBe(1)
+    expect(runGit.mock.calls.filter((call) => getGitArgs(call)[1] === '--get')).toHaveLength(2)
+  })
```
[the remark this answers is not in the document]
Comment (brennanb2025) [reply src/shared/git-config-snapshot-runner.test.ts:211]: Added in 56dedc96e: a regression that loads a snapshot and asserts an absent key rejects with `GitConfigSnapshotKeyNotFoundError`, while still being served from the loaded snapshot (asserts exactly one `config --list -z` and zero passthrough `config --get`).

**Comment by brennanb2025** [comment]
@coderabbitai review
